src/main/java/io/github/in_toto/lib/JSONEncoder.java

```diff
@@ -24,43 +25,129 @@
-    static JsonElement canonicalize(JsonElement src) {
-      if (src instanceof JsonArray) {
-        // Canonicalize each element of the array
-        JsonArray srcArray = (JsonArray)src;
-        JsonArray result = new JsonArray();
-        for (int i = 0; i < srcArray.size(); i++) {
-          result.add(canonicalize(srcArray.get(i)));
-        }
-        return result;
-      } else if (src instanceof JsonObject) {
-        // Sort the attributes by name, and the canonicalize each element of the object
-        JsonObject srcObject = (JsonObject)src;
-        JsonObject result = new JsonObject();
-        TreeSet<String> attributes = new TreeSet<>();
-        for (Map.Entry<String, JsonElement> entry : srcObject.entrySet()) {
-          attributes.add(entry.getKey());
-        }
-        for (String attribute : attributes) {
-          result.add(attribute, canonicalize(srcObject.get(attribute)));
+    /**
+     * Static helper method to canonicalize a json string by escaping the
+     * double quotes `"` and backslash `\` characters using the backslash
+     * escape character `\`, based on the securesystemslib implementation used
+     * in the in-toto reference implementation, see:
+     * https://github.com/secure-systems-lab/securesystemslib/blob/v0.11.2/securesystemslib/formats.py#L688
+     *
+     * @param src Source string to be canonicalized
+     *
+     * @return A canonicalized String
+     */
+    static String canonicalizeString(String src) {
+        String pattern = "([\\\\\"])";
+        return String.format("\"%s\"", src.replaceAll(pattern, "\\\\$1"));
+    }
+
+    /**
+     * Static helper method to recursively create a canonical json encoded
+     * string of the passed JsonElement, based on the securesystemslib
+     * implementation used in the in-toto reference implementation, see:
+     * https://github.com/secure-systems-lab/securesystemslib/blob/v0.11.2/securesystemslib/formats.py#L712
+     *
+     * @param src Source JsonElement to be traversed and encoded
+     *
+     * @return A canonical json encoded string of the passed JsonElement.
+     */
+    static String canonicalize(JsonElement src) {
+        String result =  new String();
+        if (src instanceof JsonArray) {
+            // Canonicalize each element of the array
+            result += "[";
+            JsonArray array = (JsonArray) src;
+            for (int i = 0; i < array.size(); i++) {
+                result += canonicalize(array.get(i));
+
+                if (i < array.size() - 1) {
+                  result += ",";
+                }
+            }
+            result += "]";
+
+        } else if (src instanceof JsonObject) {
+            result += "{";
+
+            JsonObject obj = (JsonObject)src;
+            // Create an ordered list of the JsonObject's keys
+            TreeSet<String> keys = new TreeSet<>();
+            for (Map.Entry<String, JsonElement> entry : obj.entrySet()) {
```

Comment:
Correct me if I'm wrong, but shouldn't these be ordered alphabetically?

Reply:
Good point. I actually reused the [code that was already there](https://github.com/in-toto/in-toto-java/pull/12/files/d84ecf62ad3ab087add0cdf3c3a92c93a9444935#diff-716698852f0323b14edbffd178b65cacL44), quickly checking the [`TreeSet` reference](https://docs.oracle.com/javase/7/docs/api/java/util/TreeSet.html#TreeSet()), which claims that elements are *"sorted according to the natural ordering of its elements"*, and thought that was okay.

Now double checking:
For `String` elements *natural ordering* means [*lexicographic ordering*](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#compareTo(java.lang.String)), which is the same what Python's `sorted` builtin does, which [we use in the reference implementation](https://github.com/secure-systems-lab/securesystemslib/blob/v0.11.2/securesystemslib/formats.py#L737). So we should be okay indeed, right?

Reply:
Yes, thanks for double checking!